### PR TITLE
[PyOV] Introduce get_data with copy/view in Constant

### PR DESCRIPTION
**Pull request by jiwaszki**
### Details:
 - Constant's function `get_data` is extended to provide both copy and view of the data.
 - `get_data` can now cast date to desired dtype.
 - Added testcases.

Example:
```python
In [1]: import openvino as ov

In [2]: import numpy as np

In [3]: arr = np.array([1.0, 1012.5, 0.0, 0.1, 0.5], dtype=np.float32)

In [4]: c = ov.runtime.opset13.constant(arr, dtype=np.float32)

In [5]: c.get_data(dtype=np.bool_)
Out[5]: array([ True,  True, False,  True,  True])

In [6]: c = ov.runtime.opset13.constant(arr, dtype=ov.Type.bf16)

In [7]: c.get_data(dtype=np.float32)
Out[7]: 
array([1.0000000e+00, 1.0120000e+03, 0.0000000e+00, 9.9609375e-02,
       5.0000000e-01], dtype=float32)

In [8]: c.get_data(dtype=np.float32, copy=False)
Out[8]: 
array([1.0000000e+00, 1.0120000e+03, 0.0000000e+00, 9.9609375e-02,
       5.0000000e-01], dtype=float32) 
```
### Tickets:
 - CVS-129930
 - CVS-127292

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/ops/constant.cpp [hunk 1]
```diff
@@ -201,33 +155,73 @@ void regclass_graph_op_Constant(py::module m) {
...
         R"(
             Access to Constant's data - creates a copy of data.
 
+            If `dtype` parameter is applied, function is casting data to
+            desired dtype.
+            If `dtype` is not specified, it's inherited from Constant itself.
+            If `copy` is enabled, `dtype` and Constant's element type must
+            match to create a view.
+            If `copy` is disabled and `dtype` and Constant's element type
+            do not match -- copy is forced by default.
```
Comment (slyalin) [inline src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:206]: And if the types do not match then we always return a view, right?

Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:206]: Let's go case by case:
```python
arr = np.array([1.0, 1012.5, 0.0, 0.1, 0.5], dtype=np.float32)
c = ov.runtime.opset13.constant(arr, dtype=ov.Type.f32)

c.get_data(dtype=np.float32, copy=False).flags["OWNDATA"]  # False -- view

c.get_data(copy=False).flags["OWNDATA"]  # False -- view, dtype is deduced from Constant

c.get_data(dtype=np.float32, copy=True).flags["OWNDATA"]   # True -- copy occurs

c.get_data(copy=True).flags["OWNDATA"]   # True -- copy occurs, dtype is deduced from Constant

c.get_data(dtype=np.int16, copy=False).flags["OWNDATA"]  # True -- copy is forced by mismatch of types

c.get_data(dtype=np.int16, copy=True).flags["OWNDATA"]   # True -- copy is forced by mismatch of types
```
Maybe docs are written in confusing way, so they need update? Do you have any guideline how to write them clearly?

File: src/bindings/python/tests/test_graph/test_constant.py [hunk 2]
```diff
@@ -533,3 +533,127 @@ def test_float_to_f8e4m3_convert(ov_type, numpy_dtype):
...
+    ],
+)
+@pytest.mark.parametrize(
+    ("copy_flag"),
+    [
+        (True),
+        (False),
+    ],
+)
+def test_get_data_casting_bf16(src_dtype, dst_dtype, copy_flag):
```
Comment (AlexKoff88) [inline src/bindings/python/tests/test_graph/test_constant.py:652]: Can we also test BF16 when no dst_dtype and copy_flag is provided?

File: src/bindings/python/tests/test_graph/test_constant.py [hunk 3]
```diff
@@ -533,3 +533,127 @@ def test_float_to_f8e4m3_convert(ov_type, numpy_dtype):
...
+        (True),
+        (False),
+    ],
+)
+def test_get_data_casting(src_dtype, dst_dtype, copy_flag):
+    data = np.random.rand(2, 4, 16) + 0.01  # do not allow 0s -- extra edge-case for bool type
+    data = data.astype(src_dtype)
+
+    ov_const = ops.constant(data, dtype=src_dtype)
+    arr = ov_const.get_data(dtype=dst_dtype, copy=copy_flag)
```
Comment (rkazants) [inline src/bindings/python/tests/test_graph/test_constant.py:583]: can we also pass OV types to `get_data` like `dtype=Type.f32`

File: src/bindings/python/src/pyopenvino/graph/ops/constant.cpp [hunk 4]
```diff
@@ -201,33 +155,73 @@ void regclass_graph_op_Constant(py::module m) {
...
             Returns numpy array with corresponding shape and dtype.
             For Constants with openvino specific element type, such as u1,
             it returns linear array, with uint8 / int8 numpy dtype.
 
+            Note: can be used to upcast BF16 data type to float32 or float64. 
+
+            :param dtype: Targeted data type.
+            :type dtype: numpy.dtype, optional, keyword-only
+            :param copy: Enable or disable copy of data.
+            :type copy: numpy.dtype, optional, keyword-only
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:217]: boolean?

File: src/bindings/python/src/pyopenvino/graph/ops/constant.cpp [hunk 5]
```diff
@@ -201,33 +155,73 @@ void regclass_graph_op_Constant(py::module m) {
...
+            do not match -- copy is forced by default.
+
             Returns numpy array with corresponding shape and dtype.
             For Constants with openvino specific element type, such as u1,
             it returns linear array, with uint8 / int8 numpy dtype.
 
+            Note: can be used to upcast BF16 data type to float32 or float64. 
+
+            :param dtype: Targeted data type.
+            :type dtype: numpy.dtype, optional, keyword-only
```
Comment (rkazants) [inline src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:215]: what about OV type?

File: src/bindings/python/src/pyopenvino/graph/ops/constant.cpp (hunk 1 above)
Comment (slyalin) [reply src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:206]: The examples make sense and reflect my understanding of how this flag should work. "Try to return a view by default, if not possible due to types mismatch or when copy=True, then make a copy", so performance is a priority, but can be overridden.

Comment (slyalin) [reply src/bindings/python/src/pyopenvino/graph/ops/constant.cpp:206]: One more example with all arguments with default values: `c.get_data()` returns a view, right? Will it fail when type is not representable in numpy, bf16?

File: src/bindings/python/tests/test_graph/test_constant.py (hunk 2 above)
Comment (jiwaszki) [reply src/bindings/python/tests/test_graph/test_constant.py:654]: Added `None` as destination type, extended test cases.

File: src/bindings/python/tests/test_graph/test_constant.py [hunk 6]
```diff
@@ -533,3 +533,127 @@ def test_float_to_f8e4m3_convert(ov_type, numpy_dtype):
...
+        (np.float16),
+        (np.float32),
+        (np.float64),
+    ],
+)
+@pytest.mark.parametrize(
+    ("dst_dtype"),
+    [
+        (np.float32),
+        (np.float64),
```
[the remark this answers is not in the document]
Comment (jiwaszki) [reply src/bindings/python/tests/test_graph/test_constant.py:644]: Done, that's for catching it!

File: src/bindings/python/tests/test_graph/test_constant.py (hunk 3 above)
Comment (jiwaszki) [reply src/bindings/python/tests/test_graph/test_constant.py:583]: It's not going to be provided (as for now). As destination types can only be of numpy type, so it won't become a case where any misunderstanding occurs.
If you want you can use `element_type.to_dtype()` to get desired type.

**Comment by jiwaszki** [comment]
@AlexKoff88 @slyalin please take a look again, thanks!

File: src/bindings/python/src/pyopenvino/graph/ops/constant.cpp (hunk 1 above)
Comment (slyalin) [reply src/bindings/python/src/pyopenvino/graph/ops/constant.cpp]: Does it make sense to hold this reinterpretation of bf16 as f16? It looks to be the only way how we can still zero-copy bf16 data and pass them to the code that relies on this reinterpretation, right?

**Review by slyalin — APPROVED** [review]

**Review by AlexKoff88 — APPROVED** [review]

**Pull request merged**